src/mpid/ch4/netmod/ofi/ofi_rma.h

```diff
@@ -384,14 +384,114 @@ static inline int MPIDI_OFI_do_put(const void *origin_addr,
     }
 
     /* noncontiguous messages */
-    if (sigreq)
-        mpi_errno =
-            MPIDIG_mpi_rput(origin_addr, origin_count, origin_datatype, target_rank, target_disp,
-                            target_count, target_datatype, win, sigreq);
-    else
-        mpi_errno =
-            MPIDIG_mpi_put(origin_addr, origin_count, origin_datatype, target_rank, target_disp,
-                           target_count, target_datatype, win);
+    MPI_Aint origin_density, target_density;
+    MPIDI_Datatype_get_density(origin_datatype, origin_density);
+    MPIDI_Datatype_get_density(target_datatype, target_density);
+
+    /* Sparse targets should unpack from temporary buffer. Let the generic AM code handle it. */
+    if (target_density < MPIR_CVAR_CH4_OFI_IOVEC_MIN_SEG_SIZE)
+        goto am_fallback;
+
+    /* target is high density, create iovecs representing target */
+    struct iovec *target_iov;
+    MPI_Aint num_iov, actual_iov_bytes;
+    int target_iov_len;
+    MPIR_Typerep_iov_len(target_count, target_datatype, target_bytes, &num_iov);
+    target_iov = MPL_malloc(sizeof(struct iovec) * num_iov, MPL_MEM_RMA);
```

Comment:
Do we need to limit the size of the IOV buffer here?  In the worst case, we'd be allocating `target_bytes / MPIR_CVAR_CH4_OFI_IOVEC_MIN_SEG_SIZE` number of IOV elements.  Is that too large?

Reply:
It could be large, yes. We could roughly allocate the number of IOVs that describe `max_msg_size` worth of data and then reuse them. That will require invoking Typerep multiple times with different offsets. Is that OK? I was originally trying to limit to a single Typerep call in each path.

Reply:
Here's the issue with IOVs: the right way to do this would be to have a limit on the memory you'd allocate.  So, for example, you can say you'll allocate a max of 16K IOV elements.  You'd get the first 16K IOV elements, issue them, get the next 16K IOV elements and so on.  This is not expensive to create.  This model is supported by yaksa, but not by Typerep.

Typerep only allows you to give offsets based on bytes, and not the number of IOV elements.  So, we'd need to calculate the number of bytes the first 16K IOV elements correspond to and then give that as an offset to Typerep.  Typerep would then convert those bytes back to IOV element count and return the next 16K elements.  This conversion is the expensive part.

One solution is to add an IOV-element-offset version to the Typerep API.  This would make yaksa take the fast path and dataloop take the slow path, which is the opposite of what we have today.  I believe such an API is more natural for what MPICH would need to use it as anyway.

Reply:
FYI, with the new API, pipelining it will not add much overhead.